### PR TITLE
[luasnip] adds support for loading lua snippets and setting extra config

### PR DESCRIPTION
**Pull request by b-m-f**
The title should describe it all.

Since I found a TODO in the code I did not open an issue first.

### CODE REVIEW

File: plugins/snippets/luasnip/default.nix [hunk 1]
```diff
@@ -88,10 +143,26 @@ in {
...
       cfg.fromVscode;
+    fromLuaLoaders =
+      lists.map
+      (
+        loader: let
+          options = attrsets.getAttrs ["paths"] loader;
+        in ''
+          require("luasnip.loaders.from_lua").${optionalString loader.lazyLoad "lazy_"}load(${helpers.toLuaObject options})
+
+        ''
```
Comment (GaetanLepage) [inline plugins/snippets/luasnip/default.nix:154]:
```suggestion
          require("luasnip.loaders.from_lua").${optionalString loader.lazyLoad "lazy_"}load(${helpers.toLuaObject options})
        ''
```

**Comment by b-m-f** [comment]
Hi @GaetanLepage,

I have included the whitespace fix. Do you want me to squash commits on my fork, or simply do this on merge?

File: plugins/snippets/luasnip/default.nix [hunk 2]
```diff
@@ -76,7 +90,48 @@ in {
...
+                  str
+                  path
+                  helpers.rawType
+                  (listOf (oneOf
+                    [
+                      str
+                      path
+                      helpers.rawType
+                    ]))
+                ]);
```
Comment (GaetanLepage) [inline plugins/snippets/luasnip/default.nix:130]:
```suggestion
          paths =
            helpers.defaultNullOpts.mkNullable
            (with types;
              nullOr (oneOf
                [
                  str
                  path
                  helpers.rawType
                  (listOf (oneOf
                    [
                      str
                      path
                      helpers.rawType
                    ]))
                ]
              )
            )
            "<UPSTREAM_DEFAULT>"
            "<DESCRIPTION>";
```

**Comment by GaetanLepage** [comment]
> I have included the whitespace fix. Do you want me to squash commits on my fork, or simply do this on merge?

As you want. You can squash now yourself, but I can also do it at merge time.
Your call :)

File: plugins/snippets/luasnip/default.nix (hunk 2 above)
Comment (traxys) [inline plugins/snippets/luasnip/default.nix:130]: to reduce a bit the verbosity I would have done
```nix
let pathTypes = [str path helpers.rawType] in nullOr (oneOf (pathTypes ++ listOf (oneOf pathTypes)))
```

You could even do `pathTypes = oneOf [....]` and use `either` but that would maybe complicate the documentation

**Comment by b-m-f** [comment]
@GaetanLepage I have update the MR and added a new commit, just in case I misunderstood sth.

@traxys Using the helper seems good for me. But I'm not yet seasoned in Nix, so if you think I should use such a **let** statement I can add it

File: plugins/snippets/luasnip/default.nix [hunk 3]
```diff
@@ -13,6 +13,20 @@ in {
 
     package = helpers.mkPackageOption "luasnip" pkgs.vimPlugins.luasnip;
 
+    extraConfig = mkOption {
+      type = types.attrs;
```
Comment (traxys) [inline plugins/snippets/luasnip/default.nix:17]: Please use `types.attrsOf types.anything`, it is better handled in recursive merges of values.

**Review by traxys — CHANGES_REQUESTED** [review]
You can leave the types like you had, it was mostly a suggestion for style I don't have much a preference.

File: plugins/snippets/luasnip/default.nix (hunk 3 above)
Comment (b-m-f) [reply plugins/snippets/luasnip/default.nix]: Updated

**Review by traxys — APPROVED** [review]

**Pull request merged**